File: apps/web/src/lib/outliers.ts
import type { OutlierAnalysis } from "./types";
import { mean, median, std } from "./stats";

// Outlier analysis that knows the difference between a SKEWED SEGMENT and a real ANOMALY.
//
// A plain |z|>3 rule flags the whole upper tail of a right-skewed column — e.g. every luxury car in a
// sales table — and then (unhelpfully, even alarmingly) says "check whether these are real". But a
// premium price tier isn't a data error. This module classifies a metric's extreme values:
//   • "skew"    — a heavy one-sided tail / distinct high (or low) segment: report it as distribution
//                 shape and steer the reader to the MEDIAN, don't cry "errors".
//   • "anomaly" — a few isolated points detached from the rest: worth checking (likely typos/glitches).
// Pure + worker-safe.

const MIN_N = 8;

/** Population skewness g1 = mean(((x−μ)/σ)³). >0 right-tailed, <0 left-tailed. */
function skewness(xs: number[], mu: number, sigma: number): number {
  if (sigma === 0 || xs.length === 0) return 0;
  let s = 0;
  for (const x of xs) s += ((x - mu) / sigma) ** 3;
  return s / xs.length;
}

/**
 * Analyze a numeric column's extreme values. Returns undefined when there's too little data, no spread,
 * or nothing extreme. `threshold` is the |z| cutoff for "extreme".
 */
export function analyzeColumnOutliers(column: string, xs: number[], threshold = 3): OutlierAnalysis | undefined {
  const finite: { index: number; value: number }[] = [];
  xs.forEach((x, i) => { if (Number.isFinite(x)) finite.push({ index: i, value: x }); });
  const n = finite.length;
  if (n < MIN_N) return undefined;

  const values = finite.map((f) => f.value);
  const mu = mean(values);
  const sigma = std(values);
  if (sigma === 0) return undefined;
  const med = median(values);
  const skew = skewness(values, mu, sigma);

  const extremes = finite
    .map((f) => ({ index: f.index, value: f.value, z: (f.value - mu) / sigma }))
    .filter((e) => Math.abs(e.z) >= threshold)
    .sort((a, b) => Math.abs(b.z) - Math.abs(a.z));
  if (!extremes.length) return undefined;

  const highCount = extremes.filter((e) => e.z > 0).length;
  const lowCount = extremes.length - highCount;
  const direction: OutlierAnalysis["direction"] = highCount && lowCount ? "both" : highCount ? "high" : "low";
  const share = extremes.length / n;
  const oneSidedFrac = Math.max(highCount, lowCount) / extremes.length;
  // How far the most-extreme value sits from the median, as a multiple. A real premium/low tier is a
  // moderate multiple (luxury cars ~5× the typical price); a 50× spike is a glitch, not a segment.
  const tailRatio = Math.abs(extremes[0].value) / (Math.abs(med) || 1);

  // It's a skewed SEGMENT (use the median, don't cry "errors") when there's a genuine one-sided TAIL —
  // several points, clearly one direction, a notably skewed column, at a plausible (not absurd) magnitude.
  // Otherwise it's isolated ANOMALIES: too few, two-sided, or so far out they read as data errors.
  const looksLikeSegment = extremes.length >= 5 && oneSidedFrac >= 0.8 && Math.abs(skew) > 1 && tailRatio <= 15;
  const kind: OutlierAnalysis["kind"] = looksLikeSegment ? "skew" : "anomaly";

  return {
    column,
    count: extremes.length,
    share,
    kind,
    direction,
    skew,
    mean: mu,
    median: med,
    std: sigma,
    examples: extremes.slice(0, 4),
    indices: extremes.map((e) => e.index),
  };
}

File: apps/pyapi/api/_outliers.py
"""Outlier analysis that knows a skewed SEGMENT from a real ANOMALY (scipy.stats).

A heavy one-sided tail (a premium price tier) is distribution SHAPE — report it and steer to the median,
don't cry "errors". A few isolated points (or an absurd 50× spike) are anomalies worth checking.

Flagging uses the ROBUST modified z-score (median + MAD), not the classic mean/std z. The mean and SD are
themselves dragged by the very outliers we're hunting (a single 50× spike inflates SD so nothing — not even
the spike — clears 3σ: the "masking" effect). The median/MAD are unmoved by a few extremes, so genuine
anomalies actually surface. We fall back to mean/SD only when MAD is 0 (a column more than half identical),
where the modified z is undefined.
"""
from __future__ import annotations

import numpy as np
from scipy import stats

# Iglewicz–Hoaglin: M_i = 0.6745·(x−median)/MAD is ~N(0,1) for normal data, so 3.5 is the standard cutoff
# (slightly stricter than the classic 3σ, which is the point — robust scores catch real outliers cleanly).
_MAD_TO_SIGMA = 0.6745
_MOD_Z_CUTOFF = 3.5


def analyze_column_outliers(name: str, values, std_threshold: float = 3.0) -> dict | None:
    x = np.asarray([v for v in values if np.isfinite(v)], dtype=float)
    n = len(x)
    if n < 8:
        return None
    mu, sigma = float(x.mean()), float(x.std(ddof=1))
    if sigma == 0:  # a constant column has no outliers
        return None
    med = float(np.median(x))
    skew = float(stats.skew(x))

    mad = float(np.median(np.abs(x - med)))
    if mad > 0:
        scores = _MAD_TO_SIGMA * (x - med) / mad  # robust modified z-score
        cutoff, scale = _MOD_Z_CUTOFF, "mad"
    else:
        # >50% of values identical → MAD is 0 and the robust score blows up; fall back to classic z.
        scores = (x - mu) / sigma
        cutoff, scale = std_threshold, "std"

    idx = np.where(np.abs(scores) >= cutoff)[0]
    if idx.size == 0:
        return None

    svals = scores[idx]
    order = idx[np.argsort(-np.abs(svals))]
    examples = [{"value": float(x[i]), "z": float(scores[i])} for i in order[:4]]
    high = int((svals > 0).sum())
    low = int(idx.size - high)
    direction = "both" if high and low else ("high" if high else "low")
    share = idx.size / n
    one_sided = max(high, low) / idx.size
    tail_ratio = abs(examples[0]["value"]) / (abs(med) or 1.0)

    looks_like_segment = idx.size >= 5 and one_sided >= 0.8 and abs(skew) > 1 and tail_ratio <= 15
    kind = "skew" if looks_like_segment else "anomaly"
    return {
        "column": name, "count": int(idx.size), "share": float(share), "kind": kind,
        "direction": direction, "skew": skew, "mean": mu, "median": med, "std": sigma,
        "scale": scale, "examples": examples,
    }
